R/decorated-future.R

```diff
@@ -0,0 +1,27 @@
+decorated_future <- function(inner, pre = function() {}, post = function() {}) {
+  structure(
+    list(
+      inner = inner,
+      pre = pre,
+      post = post
+    ),
+    class = c("DecoratedFuture", class(inner))
+  )
+}
+
+#' @export
+run.DecoratedFuture <- function(future, ...) {
+  future$pre()
+  future::run(future$inner)
+}
+
+#' @export
+resolve.DecoratedFuture <- function(x, ...) {
+  future::resolve(x$inner)
+  x$post()
```

Comment:
`x$post()` is where the power is. Like you said, it will have updates to the graph, and I think we can also throw in caching on the master process. But what about the caching on workers that @kendonB and others rely on? Would we need some post-work in `run.DecoratedFuture()` too?

Reply:
Can we do the post work in the inner future?

Reply:
Whatever we do, I think we should be consistent. If there is no post-work for `run()`, then I think there should also be no pre-work and it should be `drake`'s responsibility to supply both. This could be an elegant way to distinguish between what the worker does and what the master does. On the other hand, it requires `drake` to edit language objects, which could scale poorly with lots of targets. It may be faster to supply both `pre()` and `post()` for `run()`.